### PR TITLE
PLAT-3354: Audit vulnerable dependencies

### PR DESCRIPTION
**Pull request by talgie-ll**
Jira: [PLAT-3354](https://leaflink.atlassian.net/browse/PLAT-3354)

Updates to fix the following vulnerabilities. (From [safety](https://pypi.org/project/safety/))
Also updates black to the newest version.

```
+============================+===========+==========================+==========+
| package                    | installed | affected                 | ID       |
+============================+===========+==========================+==========+
| click                      | 7.1.2     | <8.0.0                   | 47833    |
+==============================================================================+
| Click 8.0.0 uses 'mkstemp()' instead of the deprecated & insecure            |
| 'mktemp()'.                                                                  |
| https://github.com/pallets/click/issues/1752                                 |
+==============================================================================+
| jinja2                     | 2.11.2    | <2.11.3                  | 39525    |
+==============================================================================+
| This affects the package jinja2 from 0.0.0 and before 2.11.3. The ReDoS      |
| vulnerability is mainly due to the `_punctuation_re regex` operator and its  |
| use of multiple wildcards. The last wildcard is the most exploitable as it   |
| searches for trailing punctuation. This issue can be mitigated by Markdown   |
| to format user content instead of the urlize filter, or by implementing      |
| request timeouts and limiting process memory. See: CVE-2020-28493.           |
+==============================================================================+
| babel                      | 2.8.0     | <2.9.1                   | 42203    |
+==============================================================================+
| Babel 2.9.1 includes a fix for CVE-2021-42771: Babel.Locale in Babel before  |
| 2.9.1 allows attackers to load arbitrary locale .dat files (containing       |
| serialized Python objects) via directory traversal, leading to code          |
| execution.                                                                   |
| https://github.com/python-babel/babel/pull/782                               |
+==============================================================================+
| urllib3                    | 1.25.10   | <1.26.5                  | 43975    |
+==============================================================================+
| Urllib3 1.26.5 includes a fix for CVE-2021-33503: An issue was discovered in |
| urllib3 before 1.26.5. When provided with a URL containing many @ characters |
| in the authority component, the authority regular expression exhibits        |
| catastrophic backtracking, causing a denial of service if a URL were passed  |
| as a parameter or redirected to via an HTTP redirect.                        |
| https://github.com/advisories/GHSA-q2q7-5pp4-w6pg                            |
+==============================================================================+
| py                         | 1.9.0     | <=1.9.0                  | 39253    |
+==============================================================================+
| A denial of service via regular expression in the py.path.svnwc component of |
| py (aka python-py) through 1.9.0 could be used by attackers to cause a       |
| compute-time denial of service attack by supplying malicious input to the    |
| blame functionality. See CVE-2020-29651.                                     |
+==============================================================================+
```

• Removing appdirs (1.4.4)
• Removing regex (2020.7.14)
• Removing toml (0.10.1)
• Removing typed-ast (1.4.1)
• Removing chardet (3.0.4)

• Installing charset-normalizer (2.0.12)

• click (`7.1.2` -> `8.0.4`)
• pathspec (`0.8.0` -> `0.9.0`)
• typing-extensions (`4.1.1` -> `4.2.0`)
• black (`19.10b0` -> `22.3.0`)
• py (`1.9.0` -> `1.11.0`)
• babel (`2.8.0` -> `2.10.1`)
• markupsafe (`1.1.1` -> `2.0.1`)
• jinja2 (`2.11.2` -> `3.0.3`)
• requests (`2.24.0` -> `2.27.1`)
• urllib3 (`1.25.10` -> `1.26.9`)

### CODE REVIEW

**Comment by talgie-ll** [comment]
Tests did pass locally:
```bash
tests/test_exceptions.py::test__ConcreteJsonError__not_JsonErrorArray PASSED                                                                                [  1%]
tests/test_exceptions.py::test__ConcreteJsonError__not_JsonError_list PASSED                                                                                [  3%]
tests/test_exceptions.py::test__ConcreteJsonError__from_JsonErrors PASSED                                                                                   [  5%]
tests/test_exceptions.py::test__ConcreteJsonError__from_JsonErrorArray PASSED                                                                               [  7%]
tests/test_generics.py::test___generic_error_array PASSED                                                                                                   [  9%]
tests/test_generics.py::test__generic_errors[error403-403-Forbidden] PASSED                                                                                 [ 11%]
tests/test_generics.py::test__generic_errors[error404-404-Not found] PASSED                                                                                 [ 13%]
tests/test_generics.py::test__generic_errors[error400-400-Bad Request] PASSED                                                                               [ 15%]
tests/test_models.py::test__JsonErrorSourceParameter__as_dict PASSED                                                                                        [ 17%]
tests/test_models.py::test__JsonErrorSourcePointer__as_dict PASSED                                                                                          [ 19%]
tests/test_models.py::test__JsonError____init__ PASSED                                                                                                      [ 21%]
tests/test_models.py::test__JsonError__as_dict PASSED                                                                                                       [ 23%]
tests/test_models.py::test__JsonErrorArray___unique_status_codes PASSED                                                                                     [ 25%]
tests/test_models.py::test__JsonErrorArray__status__override PASSED                                                                                         [ 27%]
tests/test_models.py::test__JsonErrorArray__status__unique_len_1 PASSED                                                                                     [ 29%]
tests/test_models.py::test__JsonErrorArray__status__unique_len_gt_1 PASSED                                                                                  [ 31%]
tests/test_models.py::test__JsonErrorArray__status__fallback PASSED                                                                                         [ 33%]
tests/test_models.py::test__JsonErrorArray__status__mixed PASSED                                                                                            [ 35%]
tests/test_models.py::test__JsonErrorArray__as_dict PASSED                                                                                                  [ 37%]
tests/test_models.py::test__JsonErrorArray__serialized PASSED                                                                                               [ 39%]
tests/test_models.py::test__JsonErrorArray____str__ PASSED                                                                                                  [ 41%]
tests/test_pyll_json_errors.py::test__check_dependency[the_package_does_not_exist-True] PASSED                                                              [ 43%]
tests/test_pyll_json_errors.py::test__check_dependency[os-False] PASSED                                                                                     [ 45%]
tests/test_pyll_json_errors.py::test_version PASSED                                                                                                         [ 47%]
tests/test_transform.py::test_BaseTransform__to_list PASSED                                                                                                 [ 49%]
tests/test_transform.py::test_BaseTransform__to_array PASSED                                                                                                [ 50%]
tests/test_utils.py::test__flatten_dict[inp0-outp0] PASSED                                                                                                  [ 52%]
tests/test_utils.py::test__flatten_dict[inp1-outp1] PASSED                                                                                                  [ 54%]
tests/contrib/test_flask.py::test__make_response PASSED                                                                                                     [ 56%]
tests/contrib/test_flask.py::test__HttpExceptionTransform__make_json_errors[exc_classes0] PASSED                                                            [ 58%]
tests/contrib/test_flask.py::test__HttpExceptionTransform__make_json_errors[exc_classes1] PASSED                                                            [ 60%]
tests/contrib/test_flask.py::test__HttpExceptionTransform__make_json_errors[exc_classes2] PASSED                                                            [ 62%]
tests/contrib/test_flask.py::test__HttpExceptionTransform__make_json_errors[exc_classes3] PASSED                                                            [ 64%]
tests/contrib/test_flask.py::test__HttpExceptionTransform__make_json_errors[exc_classes4] PASSED                                                            [ 66%]
tests/contrib/test_flask.py::test__HttpExceptionTransform__make_json_errors[exc_classes5] PASSED                                                            [ 68%]
tests/contrib/test_flask.py::test__wrap_app PASSED                                                                                                          [ 70%]
tests/contrib/test_flask.py::test__wrap_403[403] PASSED                                                                                                     [ 72%]
tests/contrib/test_flask.py::test__wrap_403[404] PASSED                                                                                                     [ 74%]
tests/contrib/test_marshmallow.py::test__ValidationErrorTransform__make_json_errors__simple PASSED                                                          [ 76%]
tests/contrib/test_marshmallow.py::test__ValidationErrorTransform__make_json_errors__nested PASSED                                                          [ 78%]
tests/contrib/test_marshmallow.py::test__ValidationErrorTransform__make_json_errors__multiple_same_field PASSED                                             [ 80%]
tests/contrib/test_marshmallow.py::test__ValidationErrorTransform__make_json_errors__numbered PASSED                                                        [ 82%]
tests/contrib/test_marshmallow.py::test__ValidationErrorTransform__make_json_errors__multiple_sources PASSED                                                [ 84%]
tests/contrib/test_marshmallow.py::test__complex_error_sample PASSED                                                                                        [ 86%]
tests/contrib/test_rest_framework.py::test__make_response PASSED                                                                                            [ 88%]
tests/contrib/test_rest_framework.py::test__drf_transform[exc0-formatted0] PASSED                                                                           [ 90%]
tests/contrib/test_rest_framework.py::test__drf_transform[exc1-formatted1] PASSED                                                                           [ 92%]
tests/contrib/test_rest_framework.py::test__drf_transform[exc2-formatted2] PASSED                                                                           [ 94%]
tests/contrib/test_rest_framework.py::test__complex_error_sample PASSED                                                                                     [ 96%]
tests/contrib/test_rest_framework.py::test__reformat_response PASSED                                                                                        [ 98%]
tests/contrib/test_rest_framework.py::test__reformat_response_not_api_exception PASSED                                                                      [100%]
```

File: pyll_json_errors/contrib/rest_framework.py [hunk 1]
```diff
@@ -147,14 +147,26 @@ def _recurse(pointer, error):
...
                 # handling a dictionary of suberrors
                 for prop, err in error.items():
-                    _recurse(pointer + [str(prop),], err)
+                    _recurse(
+                        pointer
+                        + [
+                            str(prop),
+                        ],
+                        err,
+                    )
```
Comment (shonteag) [inline pyll_json_errors/contrib/rest_framework.py:156]: This formatting doesn't seem correct to me, was this done by `black`?

Comment (talgie-ll) [reply pyll_json_errors/contrib/rest_framework.py:156]: I'll go back and see what did it.

Comment (talgie-ll) [reply pyll_json_errors/contrib/rest_framework.py]: @shonteag It looks like whenever there is a trailing comma after the last result, black (post-upgrade) wants to expand this to multiple lines. This means that things like:
```python
a = {"one": "thing",}
b = ["a list of any size",]
c = func_call(["a"] + ["trailing",], comma,)
```
will get expanded to:
```python
a = {
    "one": "thing",
}
b = [
    "a list of any size",
]
c = func_call(
    ["a"]
     + [
        "trailing",
    ],
    comma,
)
```

I was able to remove some of these unnecessary trailing commas (only required in multi-line use I believe) and now the formatting is very similar to the original. Tests still pass.

Comment (shonteag) [reply pyll_json_errors/contrib/rest_framework.py]: Yup, much cleaner, thanks for that. Trailing commas used to be needed to indicate a list when there was only one element, but python3+ did away with that I believe.

**Review by shonteag — APPROVED** [review]
Looks like these tests are run on CircleCI. We should make a tech debt ticket to address that, and shift it over to GHA.

**Review by aaronmarkey — APPROVED** [review]

**Pull request merged**